### PR TITLE
feat(model): WanDB with fabric

### LINKED ISSUE #122: [Super AI] Implement WanDB
Description:
Please Create WanDB Account and integrate to our training Code

Instructions:
- Create WanDB
- Add to our code

Deliverables:
- PR
- WanDB Account

### PR DESCRIPTION
**Pull request by MoosaTae**
## Why this PR
for logging loss and system on fabric training pipeline

## Changes
- add wandb offline mode
- change progress bar to enable only main global rank
image

## Related Issues
Close #122 

## Checklist
- [ ] PR should be in the [Naming convention](../../docs/PR_NAMING.md)
- [x] Assign yourself in to Assigneees
- [x] Tag related issues
- [x] Constants name should be ALL_CAPITAL, function name should be snake_case, and class name should be CamelCase
- [x] complex function/algorithm should have [Docstring](https://peps.python.org/pep-0257/)
- [x] 1 PR should not have more than 200 lines changes (Exception for test files). If more than that please open multiple PRs
- [x] At least PR reviewer must come from the task's team (model, eval, data)

### CODE REVIEW

**Review by boat1603 — APPROVED** [review]

File: src/model/openthaigpt_pretraining_model/lightning/utils.py [hunk 1]
```diff
@@ -96,7 +105,11 @@ def __init__(
         xformers: bool = False,
         checkpoint: bool = False,
         checkpoint_only_attention: bool = False,
+        num_nodes: int = 1,
     ):
+        if torch.cuda.get_device_name(0) == "NVIDIA A100-SXM4-40GB":
+            torch.set_float32_matmul_precision("medium")  # high
```
Comment (new5558) [inline src/model/openthaigpt_pretraining_model/lightning/utils.py:111]: Can you check is this really help with performance (Memory and Iteration per sec)

Comment (MoosaTae) [reply src/model/openthaigpt_pretraining_model/lightning/utils.py:112]: It's warning when train with A100
```
You are using a CUDA device ('NVIDIA A100-SXM4-40GB') that has Tensor Cores.  
To properly utilize them, you should set 
`torch.set_float32_matmul_precision('medium' | 'high')`  
which will trade-off precision for performance. 
For more details, read https://pytorch.org/docs/stable/generated/torch.set_float32_matmul_precision.html#torch.set_float32_matmul_precision
```
result when change from lightning (Tensor Cores and Matrix Multiplication Precision) [ref](https://lightning.ai/pages/community/tutorial/accelerating-large-language-models-with-mixed-precision-techniques/)
image

File: src/model/openthaigpt_pretraining_model/lightning/utils.py [hunk 2]
```diff
@@ -78,6 +83,10 @@ def seed_everything(seed):
     random.seed(seed)
 
 
+def compute_perplexity(loss):
```
Comment (new5558) [inline src/model/openthaigpt_pretraining_model/lightning/utils.py:86]: can you add type hint for loss

**Review by new5558 — APPROVED** [review]

File: src/model/openthaigpt_pretraining_model/lightning/utils.py (hunk 2 above)
Comment (MoosaTae) [reply src/model/openthaigpt_pretraining_model/lightning/utils.py:86]: in feat(model): Deepspeed for lightning fabric pull request #137

**Pull request merged**